Integrate FibsemMillingStage.reference_image attribute

thomasmfish opened the pull request:
Currently `stage.ref_image = ref_image` is set in `mill_stages` but `FibsemMillingStage` doesn't have a proper attribute for it, so this is a potential way of handling it more formally.

The complicating factor was the `to_dict` conversion. I have handled this by allowing `reference_image` to be `ImageSettings`, as that allows for the reference image to be retaken with the same settings. I'm not sure if that's the best way though. Another option I considered is saving the reference image and using the path, which would mean alignment could persist even if an experiment is reloaded. However, that would mean a change would be required around the saving of reference images, so I thought I'd propose this solution and see what you think.

I've also refactored out `get_stage_reference_image` to avoid code duplication between `setup_milling` and `mill_stages`.

thomasmfish committed:
FibsemMillingStage.reference_image

patrickcleeve2 reviewed (changes requested):
I get what you're going for, but I don't think this is the right approach. This reference image is only really valid at runtime as it  is meant to align between the imaging current and milling current, which is why it's re-acquired before starting milling. It's deliberately excluded from the to_dict/from_dict export for that reason. 

The helper function is a good addition, and I've made some more comments.

patrickcleeve2 commented on fibsem/milling/base.py:
```diff
@@ -97,14 +105,28 @@ def __post_init__(self):
             self.imaging.save = False
 
     def to_dict(self):
+        reference_image = None
```
Remove this

patrickcleeve2 commented on fibsem/milling/base.py:
```diff
@@ -116,14 +138,22 @@ def from_dict(cls, data: dict):
         imaging: dict = data.get("imaging", {})
         if imaging == {} or imaging.get("path", None) is None:
             imaging["path"] = None # set to None if not explicitly set
+
```
Remove this

patrickcleeve2 commented on fibsem/milling/core.py:
```diff
@@ -19,11 +19,33 @@
+            resolution=[1536, 1024],
+            beam_type=milling_stage.milling.milling_channel,
+            reduced_area=milling_stage.alignment.rect,
+            path=fcfg.DATA_CC_PATH,  # TODO: set this to the last-path?
```
This path can probably be updated to `milling_stage.imaging.path', if it's None then use the DATA_CC_PATH

patrickcleeve2 commented on fibsem/milling/base.py:
```diff
@@ -82,6 +89,7 @@ class FibsemMillingStage:
     strategy: MillingStrategy[Any] = field(default_factory=get_strategy)
     alignment: MillingAlignment = field(default_factory=MillingAlignment)
     imaging: ImageSettings = field(default_factory=ImageSettings) # settings for post-milling acquisition
+    reference_image: Optional[Union[FibsemImage, ImageSettings]] = None
```
I'm not a fan of these overloaded variables, if we want to expose the alignment imaging settings to the user it should be done via adding ImageSettings to .alignment. Otherwise it's very difficult to reason about.

patrickcleeve2 commented on fibsem/milling/core.py:
```diff
@@ -19,11 +19,33 @@
+def get_stage_reference_image(
+    microscope: FibsemMicroscope, milling_stage: FibsemMillingStage
+) -> FibsemImage:
+    ref_image = milling_stage.reference_image
```
There should only be two paths through here: 
1. return the previously acquired image
2. If image is None, acquire a new image
See comment above above overloaded variables

thomasmfish commented:
Okay, I think I've addressed all your comments. I've also moved the `get_stage_reference_image` function down into the utils section of core.py, since it makes more sense there.

thomasmfish commented:
Oh I forgot to ask - I noticed the new repo, so would you prefer I use the new one from now on?

patrickcleeve2 commented on fibsem/milling/core.py:
```diff
@@ -281,5 +269,29 @@ def acquire_images_after_milling(
+            path=path,
+            filename=f"ref_{milling_stage.name}_initial_alignment_{current_timestamp_v2()}",
+        )
+        return microscope.acquire_image(image_settings)
```
```suggestion
        image_settings = ImageSettings(
            hfw=milling_stage.milling.hfw,
            dwell_time=1e-6,
            resolution=[1536, 1024],
            beam_type=milling_stage.milling.milling_channel,
            reduced_area=milling_stage.alignment.rect,
            save=True,
            path=path,
            filename=f"ref_{milling_stage.name}_initial_alignment_{current_timestamp_v2()}",
        )
        return acquire.acquire_image(microscope, image_settings)
```
In testing this, I noticed that this doesn't actually save the reference image (previously existing bug). Needs to have save=True, and use the acquire api

patrickcleeve2 reviewed (changes requested):
Can you add the following test, as milling/test_core.py to validate the acquisitions. Other than its good.

```python
import glob
import os

import pytest

from fibsem import utils
from fibsem.milling import FibsemMillingStage, mill_stages

TEST_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "tmp-milling-test")
os.makedirs(TEST_PATH, exist_ok=True)


def clean_up():
    # remove any existing files in the test path
    for file in glob.glob(os.path.join(TEST_PATH, "*")):
        os.remove(file)
    # remove the directory if it exists
    if os.path.exists(TEST_PATH):
        os.rmdir(TEST_PATH)


def test_mill_stages_with_acquisitions():
    """Test milling stages with image acquisitions (alignment, reference images, etc.).
    Smoke test to ensure images are being acquired and saved correctly.
    """
    # setup a microscope session
    microscope, settings = utils.setup_session(manufacturer="Demo")

    milling_stage = FibsemMillingStage(name="test-stage")
    milling_stage.milling.acquire_images = True
    milling_stage.alignment.enabled = True
    milling_stage.imaging.path = TEST_PATH
    milling_stage.imaging.save = True
    mill_stages(microscope, [milling_stage])

    # check for the expected reference image
    glob_pattern = f"ref_{milling_stage.name}_initial_alignment*.tif"
    files = glob.glob(os.path.join(TEST_PATH, glob_pattern))
    assert len(files) > 0, f"No files found in {TEST_PATH} with pattern {glob_pattern}"

    # check for the beam shift alignment files
    glob_pattern = "beam_shift_alignment*.tif"
    files = glob.glob(os.path.join(TEST_PATH, glob_pattern))
    assert len(files) > 0, f"No files found in {TEST_PATH} with pattern {glob_pattern}"

    # check for the post acquisition reference image
    glob_pattern = f"ref_milling_{milling_stage.name}_finished_*.tif"
    files = glob.glob(os.path.join(TEST_PATH, glob_pattern))
    assert len(files) == 2, f"No files found in {TEST_PATH} with pattern {glob_pattern}"

    clean_up()

```

thomasmfish committed:
Ensure reference image is saved

thomasmfish committed:
Add test_mill_stages_with_acquisitions

thomasmfish commented:
> Can you add the following test, as milling/test_core.py to validate the acquisitions. Other than its good.

I've added that test and the change to ensure reference images are saved. I have made a change to the test to simplify it slightly by using pytest's built-in `tmp_path` fixture - I hope that's okay.
